noir_stdlib/src/meta/expr.nr

```diff
@@ -110,4 +110,51 @@ impl Expr {
     // docs:start:is_continue
     fn is_continue(self) -> bool {}
     // docs:end:is_continue
+
+    fn map<Env>(self, f: fn[Env](Expr) -> Option<Expr>) -> Expr {
+        let result = f(self);
+        let result = result.or_else(|| map_if(self, f));
+        let result = result.or_else(|| map_block(self, f));
+        // TODO: check the remaining possibilities...
+        // though to test this it's easier if we have `as_let`.
+        result.unwrap_or(self)
+    }
+}
+
+fn map_block<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    let exprs = expr.as_block();
+    if exprs.is_some() {
+        let exprs = exprs.unwrap();
+        let exprs = exprs.map(|expr: Expr| expr.map(f));
+        Option::some(new_block(exprs))
+    } else {
+        Option::none()
+    }
+}
+
+fn map_if<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    let if_ = expr.as_if();
+    if if_.is_some() {
+        let (condition, consequence, alternative) = if_.unwrap();
+        let condition = condition.map(f);
+        let consequence = consequence.map(f);
+        let alternative = alternative.and_then(f);
+        Option::some(new_if(condition, consequence, alternative))
+    } else {
+        Option::none()
+    }
+}
+
+fn new_block(exprs: [Expr]) -> Expr {
+    let exprs = exprs.map(|expr| quote { $expr }).join(quote { });
+    quote { { $exprs }}.as_expr().unwrap()
```

Comment:
Hmm, I wonder if mapping these to another Expr could remove the semicolons since we implicitly remove those when matching on Exprs.